api/parts/data/usage.js

```diff
@@ -268,6 +268,76 @@ usage.processLocation = function(params) {
     });
 };
 
+/**
+ * Set Location information in params but donot update it in users document
+ * @param  {params} params - params object
+ * @returns {Promise} promise which resolves upon completeing processing
+ */
+usage.setLocation = function(params) {
+    if ('tz' in params.qstring) {
+        params.user.tz = parseInt(params.qstring.tz);
+        if (isNaN(params.user.tz)) {
+            delete params.user.tz;
+        }
+    }
+
+    return new Promise(resolve => {
+        var loc = {
+            country: params.qstring.country_code,
+            city: params.qstring.city,
+            tz: params.user.tz
+        };
+
+        if ('location' in params.qstring) {
+            if (params.qstring.location) {
+                var coords = params.qstring.location.split(',');
+                if (coords.length === 2) {
+                    var lat = parseFloat(coords[0]),
+                        lon = parseFloat(coords[1]);
+
+                    if (!isNaN(lat) && !isNaN(lon)) {
+                        loc.lat = lat;
+                        loc.lon = lon;
+                    }
+                }
+            }
+        }
+
+        if (loc.lat !== undefined || (loc.country && loc.city)) {
+            locFromGeocoder(params, loc).then(loc2 => {
+                if (loc2.city && loc2.country && loc2.lat !== undefined) {
+                    usage.setUserLocation(params, loc2);
+                    return resolve();
+                }
+                else {
+                    loc2.city = loc2.country === undefined ? undefined : loc2.city;
+                    loc2.country = loc2.city === undefined ? undefined : loc2.country;
+                    locFromGeoip(loc2, params.ip_address).then(loc3 => {
+                        usage.setUserLocation(params, loc3);
+                        return resolve();
+                    });
+                }
+            });
+        }
+        else {
+            locFromGeoip(loc, params.ip_address).then(loc2 => {
+                usage.setUserLocation(params, loc2);
+                return resolve();
+            });
+        }
+    });
+};
+
+/**
+ * Set user location in params
+ * @param  {params} params - params object
+ * @param  {object} loc - location info
+ */
+usage.setUserLocation = function(params, loc) {
+    params.user.country = loc.country;
+    params.user.city = loc.city;
```

Comment:
Here I think we also have a region now in core, does it makes sense to have it here too?
https://github.com/Countly/countly-server/blob/next/api/parts/data/usage.js#L136

Reply:
So far region is not yet being used in the metrics.
Also we only use city in the metrics as it is specified as a default user prop.
It won't matter if region is present or not.
But will add region anyhow, maybe in future it could be used.